### PR TITLE
notify: make sequential notify_one calls behave like a single call.

### PR DESCRIPTION
**Pull request by ashdnazg**
## Motivation

According to the docs `notify_one()` sets a permit if there currently are no available permits and the permit should be consumed only by `notified().await`.
Before this commit, if only a single `notified().await` was called before `notify_one()`, then `notify_one()` consumed its own permit by setting the state to EMPTY. A subsequent call to `notify_one()` would then set a new permit even though the previous permit was not yet consumed by the `notified().await`.

## Solution

This commit solves the issue by transferring the responsibility of setting the state to EMPTY to the notified instead of the notifier.

Made together with @nihohit

### CODE REVIEW

File: tokio/tests/sync_notify.rs [hunk 1]
```diff
@@ -225,3 +225,17 @@ fn test_waker_update() {
...
+    let mut future = spawn(notify.notified());
+    assert_pending!(future.poll());
+
+    notify.notify_one();
+    notify.notify_one();
+
+    assert_ready!(future.poll());
+    let mut future = spawn(notify.notified());
+    assert_pending!(future.poll());
+}
```
Comment (Darksonn) [inline tokio/tests/sync_notify.rs:241]: This behavior seems incorrect to me. The first `notify_one` call notifies the future, and the second `notify_one` call stores a permit. The second future should consume that permit, which means that it should be ready when polled.

Comment (nihohit) [reply tokio/tests/sync_notify.rs:241]: What you're describing doesn't match the abstraction detailed in Notify's documentation:
"The notified().await method waits for a permit to become available and notify_one() sets a permit if there currently are no available permits [...] notified().await waits for the permit to be made available, consumes the permit, and resumes."

This means that `notify_one()` can set a permit, but `notified().await` can consume it. The concept of "The first notify_one call notifies the future" isn't part of the abstraction - only permits that are set by `notify_one()` and consumed by `notified().await`.

In the scenario detailed in the test, when described using the Notify abstraction, this happens in your interpratation of correct behavior:
1. `notified().await` waits
2. `notify_one()` sets one permit
3. `notify_one()` sets a second permit
4. `notified().await` wakes up and consumes the first permit
5. `notified().await` is called and consumes the second permit
thus breaking the invariant that only a single permit can be set.

Comment (Darksonn) [reply tokio/tests/sync_notify.rs:241]: The intended interpretation is that the future consumes the permit immediately during the call to `notify_one`, and not on its next poll.

Comment (nihohit) [reply tokio/tests/sync_notify.rs:241]: That isn't at all clear from the text - "notified().await waits for the permit to be made available, consumes the permit, and resumes." implies that these actions happen in `notified.await`, not in `notify_one()`.

Beyond that, I don't understand why the current implementation is better than the proposed one. The proposed implementation allows you to use the single permit to give stronger guarantees. For example, in this single threaded example - 

```rs
fn set(notifier: Rc<Notify>, flag: Rc<Cell<bool>>) {
  flag.set(true);
  notifier.notify_one();
}

async fn reset(notifier: Rc<Notify>, flag: Rc<Cell<bool>>) {
  notifier.notifiied.await();
  assert!(flag.replace(false));
}
```

In the existing implementation `reset` might fail the assert, because the invariant that only one waiter will be notified at a time isn't held. In my implementation the assert won't fail, because the invariant holds. What's the upside of the current implementation?

Comment (ashdnazg) [reply tokio/tests/sync_notify.rs:241]: Could I ask what the motivation is for this intended behaviour?

Also of note is that currently this behaviour only happens if  the future being polled before `notfiy_one` is called - if you remove line 233 in the test, the test passes with latest tokio.

Comment (NobodyXu) [reply tokio/tests/sync_notify.rs:241]: > Beyond that, I don't understand why the current implementation is better than the proposed one. The proposed implementation allows you to use the single permit to give stronger guarantees. For example, in this single threaded example -

That will break any crate that depends on this.

Having `notify_one` store one permit when nobody is waiting makes it possible to use `Notify` to wake up writers and flushed all buffered writes at once without worrying about accidentally missing the wakeup.

In short, it enables the users of `Notify` to wait on it while having another thread calls `notify_one` without worrying about the race condition where the waiting happens after `notify_one` and the waiter never gets waked up.

Comment (ashdnazg) [reply tokio/tests/sync_notify.rs:241]: @NobodyXu we're not changing the behaviour you're describing in this PR.

Comment (nihohit) [reply tokio/tests/sync_notify.rs:241]: @NobodyXu `notify_one` still stores one permit when nobody is waiting. See [this test](https://github.com/tokio-rs/tokio/blob/b9ae7e6659f31508fbe81e29508984d0c6f2d0b4/tokio/tests/sync_notify.rs#L53), which passes. This behavior wasn't modified by this change.

Comment (Darksonn) [reply tokio/tests/sync_notify.rs:241]: The behavior cannot be changed. That would be a breaking change. For example, the documentation includes an implementation of an mpmc channel, and your change would give introduce a race condition to this channel.
```Rust
use tokio::sync::Notify;

use std::collections::VecDeque;
use std::sync::Mutex;

struct Channel<T> {
    messages: Mutex<VecDeque<T>>,
    notify_on_sent: Notify,
}

impl<T> Channel<T> {
    pub fn send(&self, msg: T) {
        let mut locked_queue = self.messages.lock().unwrap();
        locked_queue.push_back(msg);
        drop(locked_queue);

        // Send a notification to one of the calls currently
        // waiting in a call to `recv`.
        self.notify_on_sent.notify_one();
    }

    pub fn try_recv(&self) -> Option<T> {
        let mut locked_queue = self.messages.lock().unwrap();
        locked_queue.pop_front()
    }

    pub async fn recv(&self) -> T {
        let future = self.notify_on_sent.notified();
        tokio::pin!(future);

        loop {
            // Make sure that no wakeup is lost if we get
            // `None` from `try_recv`.
            future.as_mut().enable();

            if let Some(msg) = self.try_recv() {
                return msg;
            }

            // Wait for a call to `notify_one`.
            //
            // This uses `.as_mut()` to avoid consuming the future,
            // which lets us call `Pin::set` below.
            future.as_mut().await;

            // Reset the future in case another call to
            // `try_recv` got the message before us.
            future.set(self.notify_on_sent.notified());
        }
    }
}
```
That said, I would be happy to see the documentation be improved, and to see a test for this behavior added. I did not realize that the tests were insufficient.

Comment (Darksonn) [reply tokio/tests/sync_notify.rs:241]: Hmm, I think I have misunderstood your implementation. Currently, under your implementation, you seem to have the following behavior:

 1. If there are at least two waiting futures, then one of them immediately consumes the permit. (Like what happens under the current implementation.)
 2. If there is exactly one waiting future, then that future doesn't immediately consume the permit. Instead, the permit is consumed on the next poll.

Since you preserved the existing consume-it-immediately behavior when there is more than one waiter, you haven't actually broken the mpmc channel example.

Comment (nihohit) [reply tokio/tests/sync_notify.rs:241]: Then let's clarify the behavior and verify whether this is a breaking  change :)

Do these tests satisfy you? Both of them pass before and after this change, and I think they describe your worry about a breaking change.

```rs
  #[test]
  fn test_two_notify_one_after_two_polls() {
      let notify = Notify::new();
      let mut future1 = spawn(notify.notified());
      let mut future2 = spawn(notify.notified());
      assert_pending!(future1.poll());
      assert_pending!(future2.poll());

      notify.notify_one();
      notify.notify_one();

      assert_ready!(future1.poll());
      assert_ready!(future2.poll());
  }

  #[test]
  fn test_two_notify_one_after_no_polls() {
      let notify = Notify::new();
      let mut future1 = spawn(notify.notified());
      let mut future2 = spawn(notify.notified());

      notify.notify_one();
      notify.notify_one();

      assert_ready!(future1.poll());
      assert_pending!(future2.poll());

      notify.notify_one();
      assert_ready!(future2.poll());
  }
```

Comment (Darksonn) [reply tokio/tests/sync_notify.rs:241]: > Then let's clarify the behavior

Is the behavior from [my previous comment](https://github.com/tokio-rs/tokio/pull/5305#discussion_r1054183349) what you intended? I still think its wrong, even if its different from what I originally thought you proposed.

> verify whether this is a breaking change or not

The `test_two_notify_one_after_poll` test that you added in the PR itself fails under the current implementation and succeeds under your proposed change. That makes it a breaking change, meaning that it would only be acceptable if the current behavior is a bug. However, the current behavior is consistent with what I intended.

Your two new proposed tests look correct, and are consistent with what I tried to explain when I wrote the documentation for [`Notified::enable`](https://docs.rs/tokio/latest/tokio/sync/futures/struct.Notified.html#method.enable).

Comment (nihohit) [reply tokio/tests/sync_notify.rs:241]: Well, IMO there is a bug. As I mentioned [here ](https://github.com/tokio-rs/tokio/pull/5305#discussion_r1053776048), the current implementation has a situation where it gives 2 permits, and as I mentioned [here ](https://github.com/tokio-rs/tokio/pull/5305#discussion_r1054003917), this prevents `Notify` from being used while maintaining single-threaded invariants. 
If you have the time, I'd appreciate an explanation of why you think the current behavior is correct - what scenario is it meant to enable?

Comment (nihohit) [reply tokio/tests/sync_notify.rs:241]: I'm reading the notification for `enable`, and I think that it too supports our interpration.

> This method returns true if the Notified is ready. This happens in the following situations:
> 
> 1. [..]
> 2. [..]
> 3. The future has previously been enabled or polled, and it has since then been marked ready by either consuming a permit from the Notify, or by a call to notify_one or notify_waiters that removed it from the list of futures ready to receive wakeups.

The docs there too separate the effect of `notify_one`, consuming a permit - these are distinct flows. `notify_one` wakes up a future, but the future should be the one to consume the permit, not `notify_one` - and since futures are lazy, the future shouldn't do anything until polled. In this case, "doing" is consuming the permit - it should only happen when a future is polled, after a permit was provided.

Comment (Darksonn) [reply tokio/tests/sync_notify.rs:241]: > As I mentioned [here](https://github.com/tokio-rs/tokio/pull/5305#discussion_r1053776048), the current implementation has a situation where it gives 2 permits

No, the `Notify` cannot hold two permits. The first permit is given away before the second `notify_one` call, so it never has more than one permit. Was I unclear about this earlier?

> and since futures are lazy, the future shouldn't do anything until polled

But under your implementation, the futures _do_ in fact consume permits without being polled. It's just that they only do so if they aren't the only waiting future. See [my previous comment](https://github.com/tokio-rs/tokio/pull/5305#discussion_r1054183349) where I mentioned this.

> and as I mentioned [here](https://github.com/tokio-rs/tokio/pull/5305#discussion_r1054003917), this prevents Notify from being used while maintaining single-threaded invariants

Actually, your implementation doesn't help with maintaining those invariants. Since your implementation sometimes has the old behavior, and sometimes has the new behavior, it can also trigger the asserts. You just need to make sure that you have more than one waiting future, so that you trigger the old behavior. See [this test](https://gist.github.com/rust-play/0f344af0ab2597ef23429a59882af19e) for an example that triggers the assert. (You can add it to `tests/sync_notify.rs`.)

> I'd appreciate an explanation of why you think the current behavior is correct

The current behavior was chosen because we want the behavior to be consistent, regardless of whether or not there is one or two waiting futures. Additionally, I want this test to pass:
```Rust
#[test]
fn test_two_notify_one_after_two_polls() {
    let notify = Notify::new();
    let mut future1 = spawn(notify.notified());
    let mut future2 = spawn(notify.notified());
    assert_pending!(future1.poll());
    assert_pending!(future2.poll());

    notify.notify_one();
    notify.notify_one();

    assert_ready!(future1.poll());
    assert_ready!(future2.poll());
}
```
Hypothetically, if permits were not consumed until you polled the future, then the second `notify_one` call would be a no-op, since the `Notify` already holds a permit from the first call. Then, `future1` would consume that permit when polled, and then `future2` would be pending at the end since there isn't a permit anymore.

The reason the test passes for you is that you side-step the issue by having different behavior depending on the number of waiters, which means that the behavior looks like this:
```Rust
#[test]
fn test_two_notify_one_after_two_polls() {
    let notify = Notify::new();
    let mut future1 = spawn(notify.notified());
    let mut future2 = spawn(notify.notified());
    assert_pending!(future1.poll());
    assert_pending!(future2.poll());

    notify.notify_one(); // <-- future1 consumes first permit here
    notify.notify_one();

    assert_ready!(future1.poll());
    assert_ready!(future2.poll()); // <-- future2 consumes second permit here
}
```
Which doesn't seem sensible to me.

> I'm reading the notification for enable, and I think that it too supports our interpration.

The reason that "either consuming a permit from the Notify" is mentioned in point 3 at all is so we can guarantee that the future is fused. If the first poll returns ready by consuming a permit, then a second poll will also return ready because it falls into the "either consuming a permit from the Notify" case of the third point. (I do see why this is confusing.)

Comment (nihohit) [reply tokio/tests/sync_notify.rs:241]: > You just need to make sure that you have more than one waiting future, so that you trigger the old behavior. See [this test](https://gist.github.com/rust-play/0f344af0ab2597ef23429a59882af19e) for an example that triggers the assert. (You can add it to tests/sync_notify.rs.)

Great catch, thanks for the test!
I'm sorry - I didn't understand before that your issue was with the difference between one and two or more waiters.
If we write a fix and ensure the test you wrote passe, would you consider merging it? This test would ensure this consistency.

Comment (Darksonn) [reply tokio/tests/sync_notify.rs:241]: Yes, I do have a test for the mpmc channel — I realized that I had misunderstood the change because the test passes under the current contents of your PR. You can find it [here][1].

> I didn't understand before that your issue was with the difference between one and two or more waiters. If we write a fix and ensure the test you wrote passes, would you consider merging it? This test would ensure this consistency.

I don't think you can fix the inconsistency without breaking the existing tests (or the mpmc test), but I would be happy to be proven wrong.

[1]: https://gist.github.com/rust-play/6f16727d7dd5761f7ba961a07a5f7431

**Comment by Darksonn** [comment]
Closing per the above discussion.
